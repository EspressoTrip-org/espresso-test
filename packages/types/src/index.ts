export * from './schema/definitions';
export * from './schema/kafka';
export * from './schema/api';

export * from './policy-utils';
export * from './satisfiers';
export * from './utils';
