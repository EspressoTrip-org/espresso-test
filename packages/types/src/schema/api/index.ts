export * from './policies-v1';
export * from './tokens-v1';
export * from './roles-v1';
export * from './query-v1';
export * from './users-v1';

export * from './list-resources';
export * from './assignments-v2';

export * from './routes';
