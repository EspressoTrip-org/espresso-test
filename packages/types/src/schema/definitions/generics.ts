import * as t from 'ts-codec';

export const Scoped = t.object({
  org_id: t.string
});
export type Scoped = t.Encoded<typeof Scoped>;

export const Personal = t.object({
  user_id: t.string
});
export type Personal = t.Encoded<typeof Personal>;

export const Managed = t.object({
  managed: t.boolean.optional()
});
export type Managed = t.Encoded<typeof Managed>;

export enum ActorType {
  System = 'system',
  User = 'user',
  Token = 'token'
}

export const SystemActor = t.object({
  type: t.literal(ActorType.System)
});
export type SystemActor = t.Encoded<typeof SystemActor>;

export const Actor = SystemActor.or(
  t.object({
    id: t.string,
    type: t.literal(ActorType.Token).or(t.literal(ActorType.User))
  })
);
export type Actor = t.Encoded<typeof Actor>;
