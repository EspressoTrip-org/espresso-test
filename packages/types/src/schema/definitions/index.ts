export * from './organizations';
export * from './permission';
export * from './generics';
export * from './policy';
export * from './users';
export * from './token';
export * from './role';
