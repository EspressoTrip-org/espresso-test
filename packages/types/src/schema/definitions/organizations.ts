import * as t from 'ts-codec';
import { Resource } from '@journeyapps-platform/micro-codecs';

export const Organization = t.object({
  name: t.string,
  locked: t.boolean.optional()
});
export type Organization = t.Encoded<typeof Organization>;

export const OrganizationResource = Resource.and(Organization);
export type OrganizationResource = t.Encoded<typeof OrganizationResource>;
