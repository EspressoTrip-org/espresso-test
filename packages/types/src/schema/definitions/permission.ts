import * as t from 'ts-codec';

export const StringOr = t.string.or(t.array(t.string));
export type StringOr = t.Encoded<typeof StringOr>;

export const IDResourceSelector = t.object({
  id: StringOr
});
export type IDResourceSelector = t.Encoded<typeof IDResourceSelector>;

export const SelectorLabels = t.record(StringOr);
export type SelectorLabels = t.Encoded<typeof SelectorLabels>;

export const LabelResourceSelector = t.object({
  labels: SelectorLabels
});
export type LabelResourceSelector = t.Encoded<typeof LabelResourceSelector>;

export type ResourceSelector = (IDResourceSelector | LabelResourceSelector) & {
  model: StringOr;
  parents?: ResourceSelector[];
};

export const ResourceSelector: t.Codec<ResourceSelector, ResourceSelector> = t.recursive('ResourceSelector', () => {
  return IDResourceSelector.or(LabelResourceSelector).and(
    t.object({
      model: StringOr,
      parents: t.array(ResourceSelector).optional()
    })
  );
});

export const Resource = t.object({
  scope: StringOr.optional(),
  selector: ResourceSelector
});
export type Resource = t.Encoded<typeof Resource>;

export enum PERMISSION_EFFECT {
  Allow = 'Allow',
  Deny = 'Deny'
}

export const Permission = t.object({
  effect: t.Enum(PERMISSION_EFFECT).optional(),
  action: StringOr,
  resource: Resource
});

export type Permission = t.Encoded<typeof Permission>;
