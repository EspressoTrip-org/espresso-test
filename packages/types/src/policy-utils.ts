import * as defs from './schema/definitions';

/**
 * Generates the cartesian product from a given policy statement.
 *
 * It is important that we break this up into permissions that are as singular as possible to improve the
 * effectiveness of our resolver. The cartesian product of a Policy is not the same as a Permission
 * with `$or` property values - especially when brought up in comparisons between two Policies.
 * This is because we perform 1:1 comparisons between permissions and as a result cannot
 * resolve accumulative access (access granted as a result of more than 1 permission).
 *
 * Put simply:
 *
 * :: [[a, b], [c, d]] !== [[a, d]]
 *
 * even though doing accumulative comparisons would resolve successfully (inferring a and d from two different permissions).
 *
 * This is kind of a plaster and having a smarter resolver would remove the need to do this - but it is good
 * enough for now
 */
export const createPermissionsFromStatement = (statement: defs.PolicyStatement) => {
  return statement.actions.reduce((permissions: defs.Permission[], action) => {
    return statement.resources.reduce((permissions, resource) => {
      const permission: defs.Permission = {
        action: action,
        resource: resource
      };
      if (statement.effect) {
        permission.effect = statement.effect;
      }

      permissions.push(permission);
      return permissions;
    }, permissions);
  }, []);
};

export const createPermissionsFromPolicy = (policy: defs.Policy) => {
  return policy.statements.reduce((permissions: defs.Permission[], statement) => {
    permissions.push(...createPermissionsFromStatement(statement));
    return permissions;
  }, []);
};

export const createPermissionsFromPolicies = (policies: defs.Policy[]) => {
  return policies.reduce((permissions: defs.Permission[], policy) => {
    permissions.push(...createPermissionsFromPolicy(policy));
    return permissions;
  }, []);
};
