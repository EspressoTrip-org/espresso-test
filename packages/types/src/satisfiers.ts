import * as defs from './schema/definitions';
import * as utils from './utils';

export const compareWithWildcards = (permission_field: defs.StringOr, comparison: defs.StringOr) => {
  const permission_fields = utils.castToArray(permission_field);
  const comparison_fields = utils.castToArray(comparison);
  if (permission_fields.includes('*')) {
    return true;
  }
  return utils.strictlyContains(permission_fields, comparison_fields);
};

export const compareLabels = (permission_labels: defs.SelectorLabels, comparison_labels: defs.SelectorLabels) => {
  const keys = Object.keys(permission_labels);
  if (keys.length === 0) {
    return false;
  }

  return keys.reduce((acc, key) => {
    if (!acc) {
      return acc;
    }
    if (comparison_labels[key] == null) {
      return false;
    }
    const permission_label = utils.castToArray(permission_labels[key]);
    const comparison_label = utils.castToArray(comparison_labels[key]);
    return utils.strictlyContains(permission_label, comparison_label);
  }, true);
};

export const satisfiesSelector = (
  permission_selector: defs.ResourceSelector,
  comparison_selector: defs.ResourceSelector
): boolean => {
  if (!compareWithWildcards(permission_selector.model, comparison_selector.model)) {
    return false;
  }

  if (utils.isIDResourceSelector(permission_selector)) {
    if (!utils.isIDResourceSelector(comparison_selector)) {
      // Even if the comparison selector is _not_ an ID selector, we still consider it a successful compare if the
      // permission ID selector is a wildcard. In other words:
      //
      // satisfies?({selector: {id: "*"}}, {selector: {labels: {a: "v"}}} === true
      if (!utils.castToArray(permission_selector.id).includes('*')) {
        return false;
      }
    } else {
      if (!compareWithWildcards(permission_selector.id, comparison_selector.id)) {
        return false;
      }
    }
  } else {
    if (!utils.isLabelResourceSelector(comparison_selector)) {
      return false;
    }
    if (!compareLabels(permission_selector.labels, comparison_selector.labels)) {
      return false;
    }
  }

  const length = permission_selector.parents?.length || 0;
  if (length === 0) {
    return true;
  }

  return !!permission_selector.parents?.reduce((matches: boolean, parent) => {
    if (!matches) {
      return false;
    }

    return !!comparison_selector.parents?.reduce((matches: boolean, comparison_parent) => {
      if (matches) {
        return matches;
      }
      return satisfiesSelector(parent, comparison_parent);
    }, false);
  }, true);
};

export type Satisfier = (permission: defs.Permission) => boolean;

export const permissionsSatisfy = (permissions: defs.Permission[], filters: Satisfier[], init: boolean) => {
  return utils.reduce(
    permissions,
    (satisfied, permission, reduced) => {
      if (satisfied !== init) {
        return reduced(satisfied);
      }

      return utils.reduce(
        filters,
        (satisfied, filter, reduced) => {
          if (satisfied !== init) {
            return reduced(satisfied);
          }
          return filter(permission);
        },
        init
      );
    },
    init
  );
};

/**
 * Given a set of permissions and a set of filters, return true if all filters satisfy all permissions
 */
export const allPermissionsSatisfyAllFilters = (permissions: defs.Permission[], ...filters: Satisfier[]) => {
  return permissionsSatisfy(permissions, filters, true);
};

/**
 * Given a set of permissions and a set of filters, return true if at least one filter satisfies at least
 * one permission
 */
export const onePermissionSatisfiesOneFilter = (permissions: defs.Permission[], ...filters: Satisfier[]) => {
  return permissionsSatisfy(permissions, filters, false);
};

/**
 * Given a set of filters, an action with an audience and a comparison resource return true if one of the permissions
 * in the given set match the namespaced action and resource
 */
export const satisfies = (permissions: defs.Permission[], action: defs.StringOr, comparison: defs.Resource) => {
  const filter = (permission: defs.Permission) => {
    // The actions must match
    if (!compareWithWildcards(permission.action, action)) {
      return false;
    }

    /**
     * If the comparison specifies a scope then the permission must have a scope specified and it must match
     * the comparison.
     *
     * If the comparison does not specify a scope then the permission, too, must not specify a scope
     */
    if (comparison.scope == null) {
      if (permission.resource.scope != null) {
        return false;
      }
    } else {
      if (permission.resource.scope == null) {
        return false;
      }
      if (!compareWithWildcards(permission.resource.scope, comparison.scope)) {
        return false;
      }
    }

    // The resource selector of the comparison must match the resource selector of the permission
    return satisfiesSelector(permission.resource.selector, comparison.selector);
  };

  const { allow, deny } = utils.splitByEffect(permissions);

  const positive_satisfies = onePermissionSatisfiesOneFilter(allow, filter);
  const negative_satisfies = onePermissionSatisfiesOneFilter(deny, filter);

  return positive_satisfies && !negative_satisfies;
};

export const canAssignPermission = (permissions: defs.Permission[]) => (comparison_permission: defs.Permission) => {
  return satisfies(permissions, comparison_permission.action, comparison_permission.resource);
};

export const requireScoped = (permission: defs.Permission) => {
  return !!permission.resource.scope;
};

/**
 * If the permission contains a scoped resource, the scope must match the provided value. This is
 * to ensure that permissions within scoped Policies or Roles do not grant access over other
 * scoped resources.
 *
 * We allow unscoped resources to appear in scoped policies as this is necessary for allowing
 * access to the resource denoting the scope (`organization` for example)
 */
export const requireMatchingScope = (scope: string) => (permission: defs.Permission) => {
  if (permission.resource.scope == null) {
    return true;
  }
  return permission.resource.scope === scope;
};

export const requireUnscoped = (permission: defs.Permission) => {
  return !permission.resource.scope;
};

/**
 * Creates a Satisfier filter which checks that a permission matches (with wildcards) at least one of a given
 * set of models
 */
export const requireModel =
  (...models: string[]): Satisfier =>
  (permission: defs.Permission) => {
    return models.reduce((valid: boolean, model) => {
      if (valid) {
        return valid;
      }
      return compareWithWildcards(permission.resource.selector.model, model);
    }, false);
  };

/**
 * Creates a Satisfier filter which checks that a permission matches (with wildcards) at least one of a given
 * set of actions
 */
export const requireAction =
  (...actions: string[]): Satisfier =>
  (permission: defs.Permission) => {
    return actions.reduce((valid: boolean, action) => {
      if (valid) {
        return valid;
      }
      return compareWithWildcards(permission.action, action);
    }, false);
  };

/**
 * Given a set of permissions and a set of Satisfier filters, return the subset of permissions that pass
 * all of the given permissions
 */
export const filterPermissionsBy = (permissions: defs.Permission[], ...filters: Satisfier[]) => {
  return permissions.filter((permission) => {
    return utils.reduce(
      filters,
      (_: boolean, filter, reduced) => {
        const res = filter(permission);
        if (!res) {
          return reduced(res);
        }
        return res;
      },
      true
    );
  });
};
