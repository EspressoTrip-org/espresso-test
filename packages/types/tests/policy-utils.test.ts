import { describe, test, expect } from 'vitest';

import * as policy_utils from '../src/policy-utils';
import * as defs from '../src/schema/definitions';
import * as satisfier from '../src/satisfiers';

describe('policy-utils', () => {
  test('policy deserialization', () => {
    const permissions = policy_utils.createPermissionsFromPolicies([
      {
        name: 'first',
        statements: [
          {
            actions: ['read', 'write'],
            resources: [
              {
                scope: '123',
                selector: {
                  model: 'draft',
                  id: '*',
                  parents: [
                    {
                      model: 'user',
                      id: '1'
                    }
                  ]
                }
              }
            ]
          },
          {
            actions: ['list'],
            resources: [
              {
                scope: ['123', '456'],
                selector: {
                  model: ['app', 'draft'],
                  id: '*'
                }
              }
            ]
          }
        ]
      },
      {
        name: 'second',
        statements: [
          {
            actions: ['*'],
            resources: [
              {
                scope: '*',
                selector: {
                  model: 'user',
                  id: ['1', '2']
                }
              }
            ]
          }
        ]
      },
      {
        name: 'third',
        statements: [
          {
            actions: ['read'],
            effect: defs.PERMISSION_EFFECT.Deny,
            resources: [
              {
                scope: '123',
                selector: {
                  model: 'user',
                  id: '2'
                }
              }
            ]
          }
        ]
      }
    ]);
    expect(permissions).toMatchSnapshot();
  });

  test('accumulative access should work with cartesian product', () => {
    const actor_policy: defs.Policy = {
      statements: [
        {
          actions: ['a', 'b'],
          resources: [
            {
              scope: '123',
              selector: {
                model: 'deployment',
                id: '*'
              }
            }
          ]
        },
        {
          actions: ['c', 'd'],
          resources: [
            {
              scope: '123',
              selector: {
                model: 'deployment',
                id: '*'
              }
            }
          ]
        }
      ]
    };

    const comparison_policy: defs.Policy = {
      statements: [
        {
          actions: ['a', 'd'],
          resources: [
            {
              scope: '123',
              selector: {
                model: 'deployment',
                id: '*'
              }
            }
          ]
        }
      ]
    };

    expect(
      satisfier.allPermissionsSatisfyAllFilters(
        policy_utils.createPermissionsFromPolicy(comparison_policy),
        satisfier.canAssignPermission(policy_utils.createPermissionsFromPolicy(actor_policy))
      )
    ).toBe(true);
  });
});
