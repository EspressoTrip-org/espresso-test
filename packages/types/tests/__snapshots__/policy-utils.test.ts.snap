// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`policy-utils > policy deserialization 1`] = `
[
  {
    "action": "read",
    "resource": {
      "scope": "123",
      "selector": {
        "id": "*",
        "model": "draft",
        "parents": [
          {
            "id": "1",
            "model": "user",
          },
        ],
      },
    },
  },
  {
    "action": "write",
    "resource": {
      "scope": "123",
      "selector": {
        "id": "*",
        "model": "draft",
        "parents": [
          {
            "id": "1",
            "model": "user",
          },
        ],
      },
    },
  },
  {
    "action": "list",
    "resource": {
      "scope": [
        "123",
        "456",
      ],
      "selector": {
        "id": "*",
        "model": [
          "app",
          "draft",
        ],
      },
    },
  },
  {
    "action": "*",
    "resource": {
      "scope": "*",
      "selector": {
        "id": [
          "1",
          "2",
        ],
        "model": "user",
      },
    },
  },
  {
    "action": "read",
    "effect": "Deny",
    "resource": {
      "scope": "123",
      "selector": {
        "id": "2",
        "model": "user",
      },
    },
  },
]
`;
